Align max regex trigger and alias capture groups for match and match_all

Edru2 opened the pull request:
#### Brief overview of PR changes/additions
Regex Triggers had a (non-documented?) limit of 33 capture groups for the first match but if match_all is active every successive
match had a limit of 10 capture group.
This PR aligns both to be at 33 capture group max and an error will be printed to the error console if the limit is reached.

#### Motivation for adding to Mudlet
fix #4121

#### Other info (issues closed, discussion etc)
I change the last parameter for pcre_exec to 99 as described at https://man7.org/linux/man-pages/man3/pcre_exec.3.html 
ovecsize is (or should be) a multiple of 3

Edru2 committed:
align max captures groups for match and match_all

vadi2 commented on src/TTrigger.cpp:
```diff
@@ -296,12 +296,15 @@ bool TTrigger::match_perl(char* subject, const QString& toMatch, int regexNumber
     } else if (rc == 0) {
-        qDebug() << "CRITICAL ERROR: SHOULD NOT HAPPEN->pcre_info() got wrong num of cap groups ovector only has room for %d captured substrings\n";
+        if (mpHost->mpEditorDialog) {
+            mpHost->mpEditorDialog->mpErrorConsole->print(tr("[Trigger Error:] Cannot handle matches. Too many capture groups.\n"), QColor(255, 128, 0), QColor(Qt::black));
```
```suggestion
            mpHost->mpEditorDialog->mpErrorConsole->print(tr("[Trigger Error:] 33 capture group limit exceeded, capture less groups.\n"), QColor(255, 128, 0), QColor(Qt::black));
```

smurfix replied:
ObGrammarNitpick: fewer, not less.

vadi2 replied:
Not a hard rule!

vadi2 commented on src/TTrigger.cpp:
```diff
@@ -368,7 +371,10 @@ bool TTrigger::match_perl(char* subject, const QString& toMatch, int regexNumber
         } else if (rc == 0) {
-            qDebug() << "CRITICAL ERROR: SHOULD NOT HAPPEN->pcre_info() got wrong num of cap groups ovector only has room for %d captured substrings\n";
+            if (mpHost->mpEditorDialog) {
+                mpHost->mpEditorDialog->mpErrorConsole->print(tr("[Trigger Error:] Cannot handle matches. Too many capture groups.\n"), QColor(255, 128, 0), QColor(Qt::black));
```
```suggestion
                mpHost->mpEditorDialog->mpErrorConsole->print(tr("[Trigger Error:] 33 capture group limit exceeded, capture less groups.\n"), QColor(255, 128, 0), QColor(Qt::black));
```

SlySven reviewed (commented):
Why the hell are we not using a compile time constant? Than everything can be kept synchronized even if we change the number later? :man_facepalming:

I.e. define a static const (or fall back to a `#define MAX_CAPTURE_GROUPS`) then multiple that by three to set the size of the `ovector` integer array, i.e. something like:
```
    int ovector[(3 * MAX_CAPTURE_GROUPS)]`
```

As it is I think we have space for 100 capture groups NOT 33!

vadi2 commented:
We do... but is it really efficient to define an array of 100 every time when 99.9% of the triggers are happy with 10?

Edru2 committed:
added suggestions and limited ovector

added vadis and svens suggestions

the ovector array is now limited to the max-capture-groups it was to big

SlySven commented on src/TTrigger.cpp:
```diff
@@ -294,17 +294,17 @@ bool TTrigger::match_perl(char* subject, const QString& toMatch, int regexNumber
+            mpHost->mpEditorDialog->mpErrorConsole->print(tr("[Trigger Error:] %1 capture group limit exceeded, capture less groups.\n").arg(MAX_CAPTURE_GROUPS), QColor(255, 128, 0), QColor(Qt::black));
         }
-        qDebug() << "CRITICAL ERROR: SHOULD NOT HAPPEN->pcre_info() got wrong num of cap groups ovector only has room for 33 captured substrings\n";
+        qDebug() << QStringLiteral("CRITICAL ERROR: SHOULD NOT HAPPEN->pcre_info() got wrong num of cap groups ovector only has room for %1 captured substrings").arg(MAX_CAPTURE_GROUPS);
```
The `qDebug()` method does not natively consume `QString`s IIRC - so raw literals (and a `int` value) are actually acceptable here, and since this is a serious issue (or so the "CRITICAL ERROR" is suggesting - though perhaps that is a bit OTT :man_shrugging: ) it might be worth upgrading it to a `qWarning()` so I suggest:
```cpp
    qWarning() << "CRITICAL ERROR: SHOULD NOT HAPPEN pcre_info() got wrong number of capture groups ovector only has room for" << MAX_CAPTURE_GROUPS << "captured substrings";
```

SlySven commented on src/TTrigger.cpp:
```diff
@@ -372,9 +372,9 @@ bool TTrigger::match_perl(char* subject, const QString& toMatch, int regexNumber
+                mpHost->mpEditorDialog->mpErrorConsole->print(tr("[Trigger Error:] %1 capture group limit exceeded, capture less groups.\n").arg(MAX_CAPTURE_GROUPS), QColor(255, 128, 0), QColor(Qt::black));
             }
-            qDebug() << "CRITICAL ERROR: SHOULD NOT HAPPEN->pcre_info() got wrong num of cap groups ovector only has room for 33 captured substrings\n";
+            qDebug() << QStringLiteral("CRITICAL ERROR: SHOULD NOT HAPPEN->pcre_info() got wrong num of cap groups ovector only has room for %1 captured substrings").arg(MAX_CAPTURE_GROUPS);
```
As above.

SlySven reviewed (commented):
By the way we have pretty much *the same code* - **also with a mix-up in the size passed to `pcre_exec(...)`** in the `TAlias` code (except that that is examining what the user has typed rather than what has come from the MUD Game server)!

Edru2 committed:
added suggestions + align  max_capture_groups also for alias

SlySven reviewed (approved):
LGTM :+1: 

Does this give you a :shirt: yet? :wink:

Edru2 commented:
They are still pending but it should be enough for a 👕 😉
